homeassistant/components/sensor/__init__.py

```diff
@@ -311,11 +311,24 @@ def state(self) -> Any:
             # Convert the date object to a standardized state string.
             if device_class == DEVICE_CLASS_DATE:
                 return value.date().isoformat()
-            return value.isoformat(timespec="seconds")
+
+            if value.tzinfo != timezone.utc:
+                value = value.astimezone(timezone.utc)
+
+            return value.astimezone(timezone.utc).isoformat(timespec="seconds")
 
         # Received a datetime
         if value is not None and device_class == DEVICE_CLASS_TIMESTAMP:
             try:
+                if value.tzinfo is None:  # type: ignore
```

Comment:
Instead of suppressing the typechecker, can't we use the fact that we're anyway doing `isinstance(value, (date, datetime))` above by rearranging the code?

Reply:
I had that, removed it because of review comments. `isinstance` was expensive to use. This is the result of it.

Reply:
Resolved with a suggestion you gave on Discord: casting it, added a comment to make clear why we did that.